text-utils/bits.c

```diff
@@ -0,0 +1,256 @@
+	case '&':
+		copy = *all_bits;
+		CPU_AND(all_bits, &copy, &bits);
+		break;
+	case '|':
+		copy = *all_bits;
+		CPU_OR(all_bits, &copy, &bits);
+		break;
+	case '^':
+		copy = *all_bits;
+		CPU_XOR(all_bits, &copy, &bits);
+		break;
+	case '~':
+		for (int i = 0; i < CPU_SETSIZE; i++) {
+			if (CPU_ISSET(i, &bits))
+				CPU_CLR(i, all_bits);
+		}
+		break;
+	}
+}
+
+enum output_mode {
+	MODE_BINARY,
+	MODE_GROUPED_MASK,
+	MODE_LIST,
+	MODE_MASK,
+};
+
+static void print_bits(cpu_set_t *bits, enum output_mode mode)
+{
+	bool started = false;
+	char buf[LINE_MAX];
+	ssize_t n = 0;
+
+	switch (mode) {
+	case MODE_MASK:
+		cpumask_create(buf, sizeof(buf), bits, sizeof(*bits));
+
+		/* strip leading zeroes */
+		while (buf[n] == '0')
+			n++;
+		if (buf[n] == '\0')
+			printf("0x0\n");
+		else
+			printf("0x%s\n", buf + n);
+		break;
+
+	case MODE_GROUPED_MASK:
+		cpumask_create(buf, sizeof(buf), bits, sizeof(*bits));
+
+		/* strip leading zeroes */
+		while (buf[n] == '0')
+			n++;
+
+		while (buf[n] != '\0') {
+			if (started && (n % 8) == 0)
+				printf(",");
+			if (buf[n] != '0')
+				started = true;
+			printf("%c", buf[n]);
+			n++;
+		}
+		printf("\n");
+		break;
+
+	case MODE_BINARY:
+		printf("0b");
+		for (n = CPU_SETSIZE; n >= 0; n--) {
+			if (started && ((n + 1) % 4) == 0)
+				printf("_");
+			if (CPU_ISSET(n, bits)) {
+				started = true;
+				printf("1");
+			} else if (started) {
+				printf("0");
+			}
+		}
+		printf("\n");
+		break;
+
+	case MODE_LIST:
+		cpulist_create(buf, sizeof(buf), bits, sizeof(*bits));
+		printf("%s\n", buf);
+		break;
+	}
+}
+
+static void __attribute__((__noreturn__)) usage(void)
+{
+	fputs(USAGE_HEADER, stdout);
+	fprintf(stdout, _(" %s [options] [<mask_or_list>...]\n"),
+		program_invocation_short_name);
+
+	fputs(USAGE_SEPARATOR, stdout);
+	fputsln(_("Convert bit masks from/to various formats."), stdout);
+
+	fputs(USAGE_ARGUMENTS, stdout);
+	fputsln(_(" <mask_or_list>      A set of bits specified as a hex mask value (e.g. 0xeec2)\n"
+		  "                     or as a comma-separated list of bit IDs."),
+		stdout);
+	fputs(USAGE_SEPARATOR, stdout);
+	fputsln(_("                     If not specified, arguments will be read from stdin."),
+		stdout);
+
+	fputs(USAGE_OPTIONS, stdout);
+	fprintf(stdout, USAGE_HELP_OPTIONS(21));
+
+	fputs(_("\nMode:\n"), stdout);
+	fputsln(_(" -m, --mask          Print the combined args as a hex mask value (default)."),
+		stdout);
+	fputsln(_(" -g, --grouped-mask  Print the combined args as a hex mask value in 32bit\n"
+		  "                     comma separated groups."), stdout);
+	fputsln(_(" -b, --binary        Print the combined args as a binary mask value."),
+		stdout);
+	fputsln(_(" -l, --list          Print the combined args as a compressed list of bit IDs."),
+		stdout);
+
+	fprintf(stdout, USAGE_MAN_TAIL("bits(1)"));
+	exit(EXIT_SUCCESS);
+}
+
+int main(int argc, char **argv)
+{
+	enum output_mode mode = MODE_MASK;
+	char **stdin_lines = NULL;
+	cpu_set_t bits;
```

Comment:
In the other utils (e.g., lscpu, taskset), we do not use `cpuset_t` as a variable with a fixed size, but it's always allocated according to the number of CPUs (or the number of wanted bits in the array).

The usual pattern is:
```
size_t setsize = 0;
cpu_set_t *bits = cpuset_alloc(maxbits, &setsize, NULL);
```

And the `setsize` is used rather than sizeof(bits). The advantage is that this solution is ready on an arbitrary number of bits, and it does not rely on libc hardcoded size.

What about using the same for 'bits'?  

My suggestion is also add `--maxbits <x>` command line option, where `<x>` is

 * explicitly specified number of wanted bits
 * "max"  (value from /sys/devices/system/cpu/kernel_max)
 * "possible" (value from /sys/devices/system/cpu/possible)
 * "online" (value from /sys/devices/system/cpu/online)
 
And default to "max" if `--maxbits` is not specified. 

By the way, the kernel max is 8191, and libc defaults to 1024, so I guess bits will have a hard time working on some huge systems.

Reply:
That's a good point. However, I would prefer not opening any file so that the tool may work even if `/sys` is not mounted.

I will add a `--maxbits` argument as you suggested (default to `8191` as `kernel_max`) but will only accept numeric values so that the tool can run in a vacuum :)

Is that ok?